### DISCUSSION TITLE
`xml-editor`: Add UI for establishing a `<definition>`

### DISCUSSION THREAD

**Discussion by grantfitzsimmons** [specify/specify7#5888] [Feature Requests]
https://github.com/specify/specify7/wiki/Form-System#configure-a-definition

**Not in the initial release**

We need to add a way to configure a `<definition>` in the UI so that we can use a different form definition to establish which columns are shown in grid view separate from which fields are shown when the form is expanded.

**Note:** The `<definition>` points to a `<viewdef>` name.

**Comment by maxpatiiuk** [comment]
When I last analyzed all the sp6 forms (using https://github.com/specify/specify_tools/tree/main/xml-inspector/schema/form-definitions), I was under the impression that customizing `<definition>` is never needed - maybe that was a mistaken impression.

If I remember correctly, the reasoning was that if you want a different layout, you would duplicate the `<view>`, append "Table" to the name just to differentiate it, then modify the `<viewdef>` for `type="table"` to have whatever fields you want to appear in the grid view.
Then, in subview definition, set `view="the name of the new view"`

Given the above steps, the current UI is sufficient.

@grantfitzsimmons isn't this the recommended way to make this work? if so, why is there a need to configure `<definition>`?
I remember we talked during the early form editor planning 1:1 and we came to a conclusion that altviews should not be exposed in the UI, and that user should only have way to edit the main viewdef and all other things should be changed automatically as needed

---

related doc: https://discourse.specifysoftware.org/t/edit-the-form-grid-to-display-different-fields-than-the-subform/435

---

side note: Fixing small error in the doc - https://github.com/specify/specify7/wiki/Form-System/_compare/94a0f6dd7715fe024676ad11588d966960559439...e694c27bd167de62f961bf078cfe6982ca40fe75

**Comment by grantfitzsimmons** [comment]
> Given the above steps, the current UI is sufficient.
> 
> @grantfitzsimmons isn't this the recommended way to make this work? if so, why is there a need to configure <definition>?

Right now, we are hiding the `<table name> Table` view definitions from the editor, but that's because we (mistakenly) did not know that it was necessary to have access to them.

https://discourse.specifysoftware.org/t/edit-the-form-grid-to-display-different-fields-than-the-subform/435

The `<definition>` tells Specify which view definition to use to decide which columns are shown when viewing a subview in grid (table) view.

<img width="986" alt="image" src="https://github.com/specify/specify7/assets/37256050/532eeb29-fbfb-4b44-955c-1a7c1df4b5bf">

Right now, you just see the `<viewdef>` for AccessionItems in the visual editor, but we need to either:
1. Add an option to edit the `<table name> Table` view definition as well so that you can point to another view definition to define which fields are used for the columns in grid view
or
2. Build a UI to choose a definition that decides which fields are used as columns in grid view

**Comment by maxpatiiuk** [comment]
> Build a UI to choose a definition that decides which fields are used as columns in grid view

A bit confused on this point in regards to what's already possible.

Isn't the following possible right now:
- Define a view definition, and include all the fields you wish to have in the grid view in that definition
- Add a subview to your form, set type to grid (form table) and set view to the name of the view definition created in the previous step

Now you have a form table on your form that shows whatever fields you want - where is the issue?

We could talk sync about this next week as I might be missing something - while I agree that all of this can be even better with more UI, less XML, I am still under impression that the current UI already allows accomplishing this
